pkg/util/mount/mount_linux.go

```diff
@@ -1171,3 +1211,57 @@ func doSafeOpen(pathname string, base string) (int, error) {
 
 	return finalFD, nil
 }
+
+// searchMountPoints finds all mount references to the source, returns a list of
+// mountpoints.
+// This function assumes source cannot be device.
+// Some filesystems may share a source name, e.g. tmpfs. And for bind mounting,
+// it's possible to mount a non-root path of a filesystem, so we need to use
+// root path and major:minor to represent mount source uniquely.
+// This implementation is shared between Linux and NsEnterMounter
+func searchMountPoints(hostSource, mountInfoPath string) ([]string, error) {
+	mis, err := parseMountInfo(mountInfoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	mountID := 0
+	rootPath := ""
+	majorMinor := ""
+
+	// Finding the underlying root path and major:minor if possible.
+	// We need search in backward order because it's possible for later mounts
+	// to overlap earlier mounts.
+	for i := len(mis) - 1; i >= 0; i-- {
+		if hostSource == mis[i].mountPoint {
+			// If it's a mount point.
+			mountID = mis[i].id
+			rootPath = mis[i].root
+			majorMinor = mis[i].majorMinor
+			break
+		} else if strings.HasPrefix(hostSource, mis[i].mountPoint) {
```

Comment:
Now that I see this, is `HasPrefix` safe? What if:

`source = "/mnt/foobar/local/dir"`
`mis[i].mountPoint = "/mnt/foo"`

mountPoint is prefix of source, but source does not reside on it. `pathWithinBase` might be better option. Please check. It will slow everything down -  how often is  searchMountPoints called?

(and `isShared()` is wrong too, I'll fix it)

Reply:
Not safe. Updated to use `pathWithinBase` in https://github.com/kubernetes/kubernetes/pull/62102/commits/c8007132b3d9cea50fecc63ead8c56d1c925b39f. `searchMountPoints` is called once in `<volume_plugin>.SetUpAt` , performance should be good.